src/core/smp.cc

```diff
@@ -148,4 +150,89 @@ smp_service_group_semaphore& get_smp_service_groups_semaphore(unsigned ssg_id, s
     return smp_service_groups[ssg_id].clients[t];
 }
 
+smp::smp(alien::instance& alien)
+        : _alien(alien) {
+}
+
+
+smp::~smp() = default;
+
+void
+smp::setup_prefaulter(seastar::resource::hwloc::internal::topology_holder& topo, seastar::memory::internal::numa_layout layout) {
+    // Stack guards mprotect() random pages, so the prefaulter will hard-fault.
+#ifndef SEASTAR_THREAD_STACK_GUARDS
+    _prefaulter = std::make_unique<internal::memory_prefaulter>(topo, std::move(layout));
+#endif
+}
+
+internal::memory_prefaulter::memory_prefaulter(resource::hwloc::internal::topology_holder& topo, memory::internal::numa_layout layout) {
+    for (auto& range : layout.ranges) {
+        _layout_by_node_id[range.numa_node_id].push_back(std::move(range));
+    }
+    auto page_size = getpagesize();
+    for (auto&& [numa_node_id, ranges] : _layout_by_node_id) {
+        posix_thread::attr a;
+        auto cpus_for_numa_node = resource::internal::cpus_for_numa_node_id(topo, numa_node_id);
+        if (cpus_for_numa_node) {
```

Comment:
Does it have rights to work in the "else" case at all?

Reply:
I tried to make it conservative. If it doesn't know what to do, at least don't cause harm. If we don't have a good idea where to run the prefaulter, give up.